ci: add coveralls.io integration

wolfgangwalther opened the pull request:
Wouldn't it be good to have code coverage tools in CI as well?

I have found 3 packages that send hpc coverage results to one of the ci services, but only `stack-hpc-coveralls` seems to be compatible to stack. The other 2 use cabal.

I thought it would make sense to use `stack-test` for that purpose. I'm not sure why that was specifically targeting the oldest supported PG version and changed that to the latest. Coverage results seem to be more interesting against the latest version, as that might touch more features, which are not enabled for older versions!?

To make this work, @steve-chavez, you would have to register the postgrest repo with coveralls.io. I'm not sure whether we need to add a token somewhere. The coveralls docs suggest we have to, but the source code for `stack-hpc-coveralls` tells me that the repo_token command line option is only used in case the CI is detected as `travis-ci`.

If you want to do that, I suggest you register the repo and we see what happens.

wolfgangwalther commented:
Currently failing with:

```
Error: Coveralls returned HTTP 422 Unprocessable Entity
{"message":"Couldn't find a repository matching this job.","error":true}
```

So that looks ok so far, since it's not registered, yet.

steve-chavez commented:
> Wouldn't it be good to have code coverage tools in CI as well?

@wolfgangwalther Yes! That looks interesting. I'll see about enabling coveralls.

steve-chavez commented:
I've added the repo to coveralls: https://coveralls.io/github/PostgREST/postgrest.

I'm re-running the checks to see if something happens.
However, I'm not sure ifthis is all we need. For showing coveralls on the checks, wouldn't we need a `.coveralls.yml` file?

steve-chavez commented:
> Error: Coveralls returned HTTP 422 Unprocessable Entity
{"message":"Couldn't find a repository matching this job.","error":true}

Same error. There must be something missing..

wolfgangwalther commented:
I do have coveralls set up for [vue-postgrest](https://github.com/technowledgy/vue-postgrest) without the `.coveralls.yml` file. But I remember, I had to set up that token.

I changed the setup to use some other tools: The haskell package `hpc-lcov` will convert from hpc to lcov format and then from there `node-coveralls` should take over. This does have support for the `COVERALLS_REPO_TOKEN` environment variable. You will have to set this in CircleCI to the token you can get from coveralls, citing their docs:

> The secret repo token for your repository, found at the bottom of your repository’s page on Coveralls.

I changed the base docker image for `stack-test` to `circleci/node` to support coveralls. Let's see whether everything still builds. Maybe we have to add some dependencies.

wolfgangwalther commented:
Ok, the new solution now throws the same error as before - once you create that environment variable `COVERALLS_REPO_TOKEN`, that should change though!

wolfgangwalther commented on .circleci/config.yml:
```diff
@@ -47,10 +47,11 @@ jobs:
+            sudo apt-get install -y libgmp-dev postgresql-client
             sudo apt-get install -y --only-upgrade binutils
             stack setup
+            stack install hpc-lcov
```
where and how exactly do I add this to any of the cabal or stack files instead of calling this manually here? would be important to be reflected in the cache key through that as well.

steve-chavez replied:
Hmm.. maybe it could go in `stack.yaml` -> `extra-deps`.

Then below use like `stack exec hpc-lcov && coveralls < lcov.info`?

(Not sure if it'd work)

wolfgangwalther replied:
That sounds like it makes sense. But how do I add it there? there are some hashes in there - I am not supposed to put those in by hand, am I? I tried `stack install` locally, but that didn't add anything. Any stack command that adds those as extra-deps?

steve-chavez replied:
IIRC, the hashes are only needed when stack doesn't have the package on its repos.

Looks like hpc-lcov is there though: https://www.stackage.org/lts-15.8/hoogle?q=hpc-lcov

You could try it like:
```
extra-deps:
- hpc-lcov-1.0.1
```

Actually, I've just noticed the hash is here: https://www.stackage.org/lts-16.19/package/hpc-lcov-1.0.1

```
This version can be pinned in stack with:hpc-lcov-1.0.1@sha256:f7831e13530634f0c2f4f2e613ddc15541f78ca45e4214e73f23efcfb90f68cb,2845
```

wolfgangwalther commented on .circleci/config.yml:
```diff
@@ -47,10 +47,11 @@ jobs:
             curl -L https://github.com/commercialhaskell/stack/releases/download/v2.3.1/stack-2.3.1-linux-x86_64.tar.gz | tar zx -C /tmp
             sudo mv /tmp/stack-2.3.1-linux-x86_64/stack /usr/bin
             sudo apt-get update
-            # ncat utility from nmap needed to test socket connection with curl <  7.40
```
that was left over from the io tests

steve-chavez commented:
> once you create that environment variable COVERALLS_REPO_TOKEN, that should change though!

Done!

wolfgangwalther commented:
Can you trigger a rebuild? I don't have permission to do that :)

wolfgangwalther commented:
The build currently takes ages, but that is due to caching and the changed base image I think. Therefore it has to build all deps again everytime.

wolfgangwalther commented:
I see you have a SSH session into the last build. Can you confirm the env var is correctly set? Maybe something like `coveralls -v < lcov.info` can tell you anything.

steve-chavez commented:
> I see you have a SSH session into the last build.

Sorry, somehow circle was rejecting my ssh key. But I can confirm I have the env var set up on their UI:

![2020-10-20-084428_963x70_scrot](https://user-images.githubusercontent.com/1829294/96595094-ed513b80-12b0-11eb-9cb8-5652368c43a7.png)

(If the error keeps happening I'll retry connecting)

wolfgangwalther commented:
So I had "export" in there to see all env vars and `COVERALLS_REPO_TOKEN` was not there. It should appear masked, if I understand https://circleci.com/docs/2.0/env-vars/#secrets-masking correctly. I added another `echo $COVERALLS_REPO_TOKEN`.

The `stack exec hpc-lcov` did not work. I think stack did not install the package, even though it was added in stack.yaml. Not sure what's missing.

wolfgangwalther commented:
> I added another `echo $COVERALLS_REPO_TOKEN`.

That line is just empty. Also the result of coveralls -v looks like there is no repo_token in the json that will be sent. So somehow it's not there.

steve-chavez commented:
> That line is just empty. Also the result of coveralls -v looks like there is no repo_token in the json that will be sent. So somehow it's not there.

Just managed to SSH to the instance. You're right. None of the env vars show with `echo` though(GITHUB_TOKEN, CACHIX_SIGNING_KEY, etc).

wolfgangwalther commented:
Are those maybe tied to jobs that run on a specific branch or something?

steve-chavez commented:
Yes, looks like they're disabled for PRs. See below:

![2020-10-20-120131_998x294_scrot](https://user-images.githubusercontent.com/1829294/96619632-05828400-12cc-11eb-89d3-524494a9c6f7.png)

steve-chavez commented:
I'll enable passing secrets for this PR. I'll disable afterwards.

wolfgangwalther commented:
Maybe we'll have to set up the coverage job with a different ci provider then, where the coverage token would be the only secret?

When I remember correctly, travis doesn't even need a repo token, they do it differently somehow. So maybe that would be the best way to go forward.

steve-chavez commented:
> When I remember correctly, travis doesn't even need a repo token, they do it differently somehow. So maybe that would be the best way to go forward.

Yes, that would be better from a security standpoint. Circle has secrets for deploying to docker, cachix, github token for releases, etc.

wolfgangwalther commented:
Wuhuuu - it works. Travis builds and uploads coverage without token.

wolfgangwalther commented:
Now, looking at https://coveralls.io/builds/34347135, we might need to fix a couple of small things:
- why is anything in `.stack-work` showing up in that report... that seems wrong
- `main/...` is not showing up. do we need to run the io tests to get coverage here?

wolfgangwalther commented:
Now that we're running coverage on travis, I tested `stack-hpc-coveralls` again. Works as well.

Both tools produce slightly different results on the conversion from hpc to lcov/coveralls. My understanding is, that this is because hpc coverage is not line based, but lcov/coveralls is. The source-code view on coveralls.io doesn't really help that much, when compared to the html output we get from stack directly. I compared both conversions and the conversion from `stack-hpc-coveralls` seems to make a lot more sense. Just a random example:

```
// ApiRequest.hs #176
  | otherwise = Nothing
```

In the HPC output `otherwise` is marked green here (green = Always True). That makes sense, otherwise IS always true - by definition. So this is not a problem at all.
- `stack-hpc-coveralls` marks this line as green on coveralls (not missed)
- `hpc-lcov | coveralls` marks this line as red on coveralls (missed)

And `stack-hpc-coveralls` does not show stuff from `.stack-work`, so that seems better as well. The folder is still listed in coveralls, because of the first two reports from `hpc-lcov | coveralls` - maybe you can reset the coveralls reports somehow, @steve-chavez?

wolfgangwalther commented:
The file from `.stack-work` that's showing up is `module postgrest-7.0.1-JBRavZt2lgCYI0yAKBSaw/Paths_postgrest`. It's in the hpc report already, so both tools are sending it to coveralls.

It does not have a corresponding source file in our repo, but is built together with the core app - not any dependencies.

What is this?

wolfgangwalther commented:
Removed `Paths_postgrest` from postgrest.cabal. Still builds successfully locally. Not sure if the file still shows up in the coverage report, though.

wolfgangwalther commented:
Ok, I excluded `Paths_postgrest` manually from the reports and it works fine.

I'll have a look into bringing back the IO tests for more coverage of `main/...`.

wolfgangwalther commented:
Almost there, I guess. The coverage is now higher, so the results of the io-tests are merged with the those from `stack test`. But I still don't see any files from `main/...` showing up, yet... mh.

wolfgangwalther commented on test/io-tests.sh:
```diff
@@ -43,7 +43,7 @@ pgrStart(){ postgrest $1 >/dev/null 2>/dev/null & pgrPID="$!"; }
 pgrStartStdin(){ postgrest $1 >/dev/null <<< "$2" & pgrPID="$!"; }
 pgrStarted(){ kill -0 "$pgrPID" 2>/dev/null; }
-pgrStop(){ kill "$pgrPID" 2>/dev/null; pgrPID=""; }
+pgrStop(){ kill "$pgrPID" 2>/dev/null; pgrPID=""; sleep 0.1; }
```
This should fix the "segmentation fault" problem we had with stack-test as well. Not relevant anymore, because we removed the io-tests there, just noting.

Had a similar problem. Once coverage is enabled, the coverage report file from process A is not closed yet when process B already tries to write to it. This leads to the bash shell running the io-tests file erroring out with segmentation fault. In the case of stack-test I mostly observed it when the port process B tried to liste on was not closed yet from process A.

wolfgangwalther commented on nix/README.md:
```diff
@@ -168,7 +168,7 @@ give you some more background and details on how it works.
 ### `default.nix`
 
-[`default.nix`](../default.nix) is our 'respository expression' that pulls all
+[`default.nix`](../default.nix) is our 'repository expression' that pulls all
```
Just two random typos I ran across and didn't feel like opening up another PR for.

wolfgangwalther commented on test/io-tests.sh:
```diff
@@ -273,13 +273,10 @@ replaceConfigValue(){
 
 getSocketStatus() {
-  curl -sL -w "%{http_code}\\n" -o /dev/null localhost:54321
+  curl -sL -w "%{http_code}\\n" -o /dev/null --unix-socket /tmp/postgrest.sock http://localhost/
```
this will now fail with curl < 7.40, which does not support --unix-socket.. but is 5 years old. Should be fine, I guess.

wolfgangwalther commented on test/io-tests.sh:
```diff
@@ -273,13 +273,10 @@ replaceConfigValue(){
 socketConnection(){
-  # map port 54321 traffic to unix socket as workaround for curl below 7.40
-  # not supporting --unix-socket flag
-  ncat -vlk 54321 -c 'ncat -U /tmp/postgrest.sock' &
```
removed this because travis blocks the nmap package.

wolfgangwalther commented:
Done!

steve-chavez commented:
> Paths_postgrest.hs.
> What is this?

Looks like that file contains cabal metadata, which we use to output the version number: https://stackoverflow.com/questions/2892586/how-can-my-haskell-program-or-library-find-its-version-number.

> Done!

@wolfgangwalther Awesome work here!

Merging now.. :rocket:
